File: native/android/app/src/main/java/io/github/paulhammant/liveverifyprototype/core/VerifyLineParser.kt
package io.github.paulhammant.liveverifyprototype.core

import java.net.URI

object VerifyLineParser {
    data class ExtractedUrl(
        val url: String,
        val urlLineIndex: Int,
    )

    fun extractVerificationUrl(rawText: String): ExtractedUrl? {
        val rawLines = rawText.split("\n").map { it.trim() }

        for (i in rawLines.indices.reversed()) {
            val line = rawLines[i]
            if (line.isBlank()) continue

            val lineNoSpaces = line.replace(Regex("\\s+"), "")
            val lower = lineNoSpaces.lowercase()

            val isCorruptedPrefix =
                lower.startsWith("verity:") ||
                    lower.startsWith("vty:") ||
                    lower.startsWith("verily:") ||
                    lower.startsWith("veryfy:")

            if (
                lower.startsWith("verify:") ||
                lower.startsWith("vfy:") ||
                isCorruptedPrefix
            ) {
                var normalized = lineNoSpaces
                if (isCorruptedPrefix) {
                    val colonIndex = normalized.indexOf(':')
                    if (colonIndex >= 0 && colonIndex + 1 < normalized.length) {
                        normalized = "vfy:" + normalized.substring(colonIndex + 1)
                    }
                }
                return ExtractedUrl(url = normalized, urlLineIndex = i)
            }
        }

        return null
    }

    fun extractCertText(rawText: String, urlLineIndex: Int): String {
        val rawLines = rawText.split("\n").map { it.trim() }
        if (urlLineIndex < 0) return rawText

        val certLines = rawLines.take(urlLineIndex).toMutableList()
        while (certLines.isNotEmpty() && certLines.last().isBlank()) {
            certLines.removeAt(certLines.lastIndex)
        }
        return certLines.joinToString("\n")
    }

    fun httpsBaseUrlString(baseUrl: String): String {
        val lower = baseUrl.lowercase()
        return when {
            lower.startsWith("verify:") -> "https://" + baseUrl.substring("verify:".length)
            lower.startsWith("vfy:") -> "https://" + baseUrl.substring("vfy:".length)
            else -> throw IllegalArgumentException("Invalid base URL format. Must start with verify: or vfy:")
        }
    }

    fun buildVerificationUrl(baseUrl: String, hash: String): URI {
        val httpsBase = httpsBaseUrlString(baseUrl)
        return URI.create("$httpsBase/$hash")
    }

    fun buildMetaUrl(baseUrl: String): URI {
        val httpsBase = httpsBaseUrlString(baseUrl)
        return URI.create("$httpsBase/verification-meta.json")
    }
}

File: native/ios/LiveVerifyPrototype/LiveVerifyPrototype/Pipeline/VerifyLineParser.swift
import Foundation

enum VerifyLineParser {
    struct ExtractedURL {
        let url: String
        let urlLineIndex: Int
    }

    static func extractVerificationURL(from rawText: String) -> ExtractedURL? {
        let rawLines = rawText.split(separator: "\n", omittingEmptySubsequences: false).map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }

        for index in rawLines.indices.reversed() {
            let line = rawLines[index]
            if line.isEmpty { continue }

            let lineNoSpaces = line.replacingOccurrences(of: #"\s+"#, with: "", options: .regularExpression)
            let lower = lineNoSpaces.lowercased()

            let prefixes = [
                "verify:",
                "vfy:",
                "verity:",
                "vty:",
                "verily:",
                "veryfy:",
            ]

            guard prefixes.contains(where: { lower.hasPrefix($0) }) else { continue }

            var normalized = lineNoSpaces
            if lower.hasPrefix("verity:") || lower.hasPrefix("vty:") || lower.hasPrefix("verily:") || lower.hasPrefix("veryfy:") {
                if let colonIndex = normalized.firstIndex(of: ":") {
                    let afterColon = normalized[normalized.index(after: colonIndex)...]
                    normalized = "vfy:" + afterColon
                }
            }

            return ExtractedURL(url: normalized, urlLineIndex: index)
        }

        return nil
    }

    static func extractCertText(from rawText: String, urlLineIndex: Int) -> String {
        let rawLines = rawText.split(separator: "\n", omittingEmptySubsequences: false).map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
        guard urlLineIndex >= 0 else { return rawText }

        var certLines = Array(rawLines.prefix(urlLineIndex))
        while let last = certLines.last, last.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            certLines.removeLast()
        }
        return certLines.joined(separator: "\n")
    }

    static func httpsBaseURLString(from baseURL: String) throws -> String {
        let lower = baseURL.lowercased()
        if lower.hasPrefix("verify:") {
            return "https://" + String(baseURL.dropFirst("verify:".count))
        }
        if lower.hasPrefix("vfy:") {
            return "https://" + String(baseURL.dropFirst("vfy:".count))
        }
        throw LiveVerifyError.invalidBaseURL
    }

    static func buildVerificationURL(baseURL: String, hash: String) throws -> URL {
        let httpsBase = try httpsBaseURLString(from: baseURL)
        guard let url = URL(string: httpsBase + "/" + hash) else {
            throw LiveVerifyError.invalidVerificationURL
        }
        return url
    }

    static func buildMetaURL(baseURL: String) throws -> URL {
        let httpsBase = try httpsBaseURLString(from: baseURL)
        guard let url = URL(string: httpsBase + "/verification-meta.json") else {
            throw LiveVerifyError.invalidBaseURL
        }
        return url
    }
}
